pkg/querier/queryrange/downstreamer.go

```
@@ -0,0 +1,185 @@
+type instance struct {
+	parallelism int
+	locks       chan struct{}
+	handler     queryrange.Handler
+}
+
+func (i instance) Downstream(ctx context.Context, queries []logql.DownstreamQuery) ([]logql.Result, error) {
+	return i.For(queries, func(qry logql.DownstreamQuery) (logql.Result, error) {
+		req := ParamsToLokiRequest(qry.Params).WithShards(qry.Shards).WithQuery(qry.Expr.String()).(*LokiRequest)
+		logger, ctx := spanlogger.New(ctx, "DownstreamHandler.instance")
+		defer logger.Finish()
+		level.Debug(logger).Log("shards", req.Shards, "query", req.Query)
+
+		res, err := i.handler.Do(ctx, req)
+		if err != nil {
+			return logql.Result{}, err
+		}
+		return ResponseToResult(res)
+	})
+}
+
+// For runs a function against a list of queries, collecting the results or returning an error. The indices are preserved such that input[i] maps to output[i].
+func (in instance) For(
+	queries []logql.DownstreamQuery,
+	fn func(logql.DownstreamQuery) (logql.Result, error),
+) ([]logql.Result, error) {
+	type resp struct {
+		i   int
+		res logql.Result
+		err error
+	}
+
+	done := make(chan struct{})
+	defer close(done)
+
+	ch := make(chan resp)
+
+	// Make one goroutine to dispatch the other goroutines, bounded by instance parallelism
+	go func() {
+		for i := 0; i < len(queries); i++ {
+			select {
+			case <-done:
+				break
+			case <-in.locks:
+				go func(i int) {
+					// release lock back into pool
+					defer func() {
+						in.locks <- struct{}{}
+					}()
+
+					res, err := fn(queries[i])
+					response := resp{
+						i:   i,
+						res: res,
+						err: err,
+					}
+
+					// Feed the result into the channel unless the work has completed.
+					select {
+					case <-done:
+					case ch <- response:
+					}
+				}(i)
+			}
+		}
+	}()
+
+	results := make([]logql.Result, len(queries))
+	for i := 0; i < len(queries); i++ {
+		resp := <-ch
+		if resp.err != nil {
+			return nil, resp.err
+		}
+		results[resp.i] = resp.res
+	}
```

A lot of code depends on the fact that len(results) == len(queries). I suggest you have a check here, in case it's not equal you could return an error.

This way we avoid users to panic on out of array bounds.

---

Actually I didn't read the code but that for does exactly that.